harness/determined/common/api/_session.py

```diff
@@ -170,6 +172,12 @@ def put(
     ) -> requests.Response:
         return self._do_request("PUT", path, params, json, data, headers, timeout, False)
 
+    def with_retry(self: T, max_retries: GeneralizedRetry) -> T:
```

Comment:
nit: the generic types are hard to read IMO. `def with_retry(self, max_retries: GeneralizedRetry) -> "BaseSession":` feels much more straightforward to me

Reply:
With the generic types:
```python
UnauthSession(...).with_retry(...)  # type = UnauthSession
Session.with_retry(...)  # type = Session
```

Without:
```python
UnauthSession(...).with_retry(...)  # type = BaseSession
Session.with_retry(...)  # type = BaseSession
```

In the latter case, someone would have to cast a `Session` from a `BaseSession` back to a `Session` (or `UnauthSession`) if they want the type checker to understand what it is.